File: 2022/elixir/aoc_2022/lib/day_1/day1.ex
defmodule Day1 do
  @inputs ["lib/day_1/test.txt", "lib/day_1/day1input.txt"]

  def solution() do
    @inputs
    |> Enum.map(&solution_helper/1)
  end

  def solution_helper(input_file) do
    input =
      input_file
      |> File.read!()
      |> String.split("\n\n")

    {{"Part 1 using '#{input_file}'", part1(input)},
     {"Part 2 using '#{input_file}'", part2(input)}}
  end

  def part1(input) do
    input
    |> Enum.reduce(0, &(&1 |> transform_to_sum() |> max(&2)))
  end

  def part2(input) do
    input
    |> Enum.reduce({0, 0, 0}, fn elem, acc ->
      cals = elem |> transform_to_sum()

      cond do
        cals > elem(acc, 0) -> {cals, elem(acc, 0), elem(acc, 1)}
        cals > elem(acc, 1) -> {elem(acc, 0), cals, elem(acc, 1)}
        cals > elem(acc, 2) -> {elem(acc, 0), elem(acc, 1), cals}
        true -> acc
      end
    end)
    |> Tuple.sum()

    # alternative lazy method with sorting and taking the first 3
    # input
    # |> Enum.map(&transform_to_sum/1)
    # |> Enum.sort(&(&1 >= &2))
    # |> Enum.take(3)
    # |> Enum.sum()
  end

  defp transform_to_sum(string) do
    string
    |> String.split("\n", trim: true)
    |> Enum.map(&String.to_integer/1)
    |> Enum.sum()
  end
end

File: 2022/rust/aoc/src/bin/day1.rs
const INPUTS: [&str; 2] = ["./inputs/1.test", "./inputs/1.input"];

fn part2(input: &String) -> i64 {
    let sums = input
        .split("\n\n")
        .map(|e| e.lines().map(|n| n.parse::<i64>().unwrap()).sum::<i64>());
    let mut max1 = 0;
    let mut max2 = 0;
    let mut max3 = 0;
    for sum in sums {
        if sum > max1 {
            max3 = max2;
            max2 = max1;
            max1 = sum;
        } else if sum > max2 {
            max3 = max2;
            max2 = sum;
        } else if sum > max3 {
            max3 = sum;
        }
    }

    max1 + max2 + max3
}

fn part1(input: &String) -> i64 {
    input.split("\n\n").fold(0, |acc, e| {
        std::cmp::max(
            e.lines().map(|n| n.parse::<i64>().unwrap()).sum::<i64>(),
            acc,
        )
    })
}

fn solution_helper(input_file: &'static str) -> (i64, i64) {
    let input = std::fs::read_to_string(input_file).expect("Unable to open file!");

    (part1(&input), part2(&input))
}

fn main() {
    for input in INPUTS {
        println!("{:?}", solution_helper(input));
    }
}
